[IR] Add nofreeobj parameter/return attribute (#206445)
llvm/llvm-project: llvm/docs/LangRef.rst

    This is not a valid attribute for return values.

``nofreeobj``

>     On arguments, this indicates that the underlying object of the argument

Reviewer: Should we mention explicitly nofreeobj strictly implies nofree for arguments?

Reviewer 2: Strictly speaking, `nofree` changes the provenance of the pointer while `nofreeobj` does not, which could make a difference for code that does silly things like take the first 4 bytes of pointer A and the last 4 bytes of pointer B and concatenates them and turns them in a new pointer...

So while ideally we'd like `nofreeobj` to imply `nofree`, I am not sure how to obtain a precise spec that actually achieves that.

Reviewer: Sorry, maybe I'm missing something. Say both argument pointer A and B carry nofree (not strictly required, but yeah). I thought we had said `inttoptr(concat_ptr)` could angelically choose any previously exposed provenance (including the original unrestricted provenance of A, the one belonging to the caller, assuming it was previously exposed), meaning inttoptr could pick a provenance that lacks `nofree`. This never happens with `nofreeobj`, where freeing the underlying object is always forbidden.

Thus, isn't `concat_ptr` more constrained under nofreeobj than under nofree (i.e., nofreeobj is a (much) stronger guarantee that would entail the (weaker) nofree)?

Reviewer 2: My example did not involve and inttoptr. I should also note that the example is very nitpicky (and @nikic is aware of the example to my knowledge), though it does reflect a real issue in the provenance model and in the past, such issues have eventually come to bite us even if they started out entirely theoretical. Also, tools like Alive have to implement *some* provenance model, so if we disregard the model then we're taking away the foundation for such tools.

I'll demonstrate it with some Rust code as that's the syntax I am most familiar with:
```rust
fn mix_pointers<T>(a: *mut T, b: *mut T) -> *mut T {
  assert_eq!(a.addr(), b.addr());
  assert_eq!(mem::size_of::<*mut T>(), 8);

  let mut result: *mut T = ptr::null_mut();

  let a_bytes = &a as *const *mut T as *const MaybeUninit<u8>;
  let b_bytes = &b as *const *mut T as *const MaybeUninit<u8>;
  let res_bytes = &mut result as *mut *mut T as *mut MaybeUninit<u8>;

  unsafe {
    res_bytes.copy_from(a_bytes, 4);
    res_bytes.add(4).copy_from(b_bytes.add(4), 4);
  }

  return result;
}
```
This function takes two pointers `a` and `b` with the same address but potentially different provenance, and returns a new pointer that mixes the provenance of `a` and `b` across its bytes. It's basically "reserve some space for the output; memcpy the first 4 bytes of `a` to the first 4 bytes of the output; memcpy the last 4 bytes of `b` to the last 4 bytes of the output". If `a` and `b` have the same provenance, the result also has that provenance and all is good, but if `a` and `b` have different provenance then the result pointer will be a no-provenance pointer because encountering mixed provenance on a `ptr`-typed load (the load we are implicilty doing at the end when we take the output buffer and make it a pointer) discards the provenance.

`nofree` changes the provenance, so if we have a function with a `ptr nofree %a` and `ptr %b` and pass the same pointer twice, and then inside we function we call `mix_pointers` on `%a` and `%b`, the result will be a no-provenance pointer. Loading from that pointer is UB. But if we had `ptr nofreeobj %a` then the mixed result has the provenance of `a` and `b` (which is the same) and there would be no UB. Therefore, it is strictly speaking not correct to say that `nofreeobj` implies `nofree`.

Reviewer: Oh, I see that now, admittedly hadn't thought about loading the pointer back from memory, which would now yield nullary provenance as the bytes have been mixed (thus leading to UB when accessing memory through it, which wouldn't happen w/ nofreeobj as it doesn't change any pointer provenance). Thanks for the elucidation!

    cannot be freed during the execution of the function (where "during" is in
    the same sense as for ``nofree``).

    On return values, this indicates that the underlying object of the return

>     value cannot be freed at all after this point.

Reviewer: Should you like this better:
```suggestion
    value cannot be freed from this point forward, throughout program's execution.
```

    Unlike ``nofree``, it is not possible to free the underlying object through

>     a different pointer either.

Reviewer: Following up on https://github.com/llvm/llvm-project/pull/195658#issuecomment-4386465882, should we be explicit and mention the interaction with noalias (nofreeobj being primarily useful in cases noalias does not apply)?

Author: I added some discussion on the relationship between nofreeobj and nofree/noalias.
